File: ios/Model/PlayerUpdateOptions.swift
import Foundation

/**
 * Update options for the TrackPlayer that can be changed at runtime.
 * These options control player behavior and capabilities that can be modified during playback.
 */
public class PlayerUpdateOptions {
  // MARK: - Core Properties

  /// Jump intervals
  public var forwardJumpInterval: Double = 15.0
  public var backwardJumpInterval: Double = 15.0
  public var progressUpdateEventInterval: Double?

  /// Rating and capabilities
  public var capabilities: [Capability] = [
    .play,
    .pause,
    .next,
    .previous,
    .seek,
  ]

  /// Repeat mode
  public var repeatMode: RepeatMode = .off

  /// iOS-specific options
  public var likeOptions: FeedbackOptions = .init(title: "Like")
  public var dislikeOptions: FeedbackOptions = .init(title: "Dislike")
  public var bookmarkOptions: FeedbackOptions = .init(title: "Bookmark")

  // MARK: - Initialization

  public init() {}

  // MARK: - Bridge Methods

  public func updateFromBridge(_ options: [String: Any]) {
    // Update jump intervals
    if let interval = options["forwardJumpInterval"] as? NSNumber {
      forwardJumpInterval = interval.doubleValue
    }
    if let interval = options["backwardJumpInterval"] as? NSNumber {
      backwardJumpInterval = interval.doubleValue
    }
    if options.keys.contains("progressUpdateEventInterval") {
      if options["progressUpdateEventInterval"] is NSNull {
        progressUpdateEventInterval = nil
      } else if let interval = options["progressUpdateEventInterval"] as? NSNumber {
        progressUpdateEventInterval = interval.doubleValue
      }
    }

    // Update capabilities
    if let caps = options["capabilities"] as? [String] {
      var updatedCapabilities = caps.compactMap { Capability(rawValue: $0) }
      // Add toggle-play-pause if both play and pause are present
      if updatedCapabilities.contains(.play), updatedCapabilities.contains(.pause) {
        updatedCapabilities.append(.togglePlayPause)
      }
      capabilities = updatedCapabilities
    }

    // Update repeat mode
    if options.keys.contains("repeatMode") {
      if options["repeatMode"] is NSNull {
        repeatMode = .off // Reset to default when explicitly set to null
      } else if let modeString = options["repeatMode"] as? String {
        repeatMode = RepeatMode(rawValue: modeString) ?? .off
      }
    }

    // Update iOS-specific options
    if let iosOptions = options["ios"] as? [String: Any] {
      if let likeDict = iosOptions["likeOptions"] as? [String: Any],
         let like = FeedbackOptions.fromBridge(likeDict)
      {
        likeOptions = like
      }
      if let dislikeDict = iosOptions["dislikeOptions"] as? [String: Any],
         let dislike = FeedbackOptions.fromBridge(dislikeDict)
      {
        dislikeOptions = dislike
      }
      if let bookmarkDict = iosOptions["bookmarkOptions"] as? [String: Any],
         let bookmark = FeedbackOptions.fromBridge(bookmarkDict)
      {
        bookmarkOptions = bookmark
      }
    }
  }

  public func toBridge() -> [String: Any] {
    var result: [String: Any] = [:]

    // Add jump intervals (always include these core values)
    result["forwardJumpInterval"] = NSNumber(value: forwardJumpInterval)
    result["backwardJumpInterval"] = NSNumber(value: backwardJumpInterval)

    // Add progress update interval (always include, nil means disabled)
    if let interval = progressUpdateEventInterval {
      result["progressUpdateEventInterval"] = NSNumber(value: interval)
    } else {
      result["progressUpdateEventInterval"] = NSNull()
    }

    // Add capabilities (always include, filter out auto-added toggle-play-pause)
    let filteredCapabilities = capabilities.filter { $0 != .togglePlayPause }.map(\.rawValue)
    result["capabilities"] = filteredCapabilities

    // Add repeat mode (always include)
    result["repeatMode"] = repeatMode.rawValue

    // Add iOS-specific options (always include with defaults)
    result["ios"] = [
      "likeOptions": likeOptions.toBridge(),
      "dislikeOptions": dislikeOptions.toBridge(),
      "bookmarkOptions": bookmarkOptions.toBridge(),
    ]

    return result
  }

  // MARK: - Convenience Methods
}

File: android/src/main/java/com/doublesymmetry/trackplayer/model/PlayerUpdateOptions.kt
package com.doublesymmetry.trackplayer.model

import com.doublesymmetry.trackplayer.option.PlayerCapability
import com.doublesymmetry.trackplayer.option.RepeatMode
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableMap

/**
 * Update options for the TrackPlayer that can be changed at runtime. These options control player
 * behavior and capabilities that can be modified during playback.
 */
data class PlayerUpdateOptions(
  // Jump intervals
  var forwardJumpInterval: Double = 15.0,
  var backwardJumpInterval: Double = 15.0,
  var progressUpdateEventInterval: Double? = null,

  // Rating and capabilities
  var capabilities: List<PlayerCapability> =
    listOf(
      PlayerCapability.PLAY,
      PlayerCapability.PAUSE,
      PlayerCapability.SKIP_TO_NEXT,
      PlayerCapability.SKIP_TO_PREVIOUS,
    ),
  var notificationCapabilities: List<PlayerCapability>? = null,

  // Repeat mode
  var repeatMode: RepeatMode = RepeatMode.OFF,

  // Android-specific runtime options (all under android.* in JS)
  var ratingType: RatingType? = null,
  var appKilledPlaybackBehavior: String? = null,
  var skipSilence: Boolean? = null,
  var shuffle: Boolean? = null,
) {

  fun updateFromBridge(map: ReadableMap?) {
    if (map == null) return

    if (map.hasKey("forwardJumpInterval")) {
      forwardJumpInterval = map.getDouble("forwardJumpInterval")
    }
    if (map.hasKey("backwardJumpInterval")) {
      backwardJumpInterval = map.getDouble("backwardJumpInterval")
    }
    if (map.hasKey("progressUpdateEventInterval")) {
      progressUpdateEventInterval =
        if (map.isNull("progressUpdateEventInterval")) {
          null
        } else {
          map.getDouble("progressUpdateEventInterval")
        }
    }

    map.getArray("capabilities")?.let { arr ->
      capabilities =
        (0 until arr.size()).mapNotNull { index ->
          val value = arr.getString(index) ?: return@mapNotNull null
          PlayerCapability.fromString(value)
            ?: throw IllegalArgumentException("Invalid capability value: $value")
        }
    }

    if (map.hasKey("notificationCapabilities")) {
      notificationCapabilities =
        if (map.isNull("notificationCapabilities")) {
          null // Explicitly set to null - reset to default behavior
        } else {
          map.getArray("notificationCapabilities")?.let { arr ->
            (0 until arr.size()).mapNotNull { index ->
              val value = arr.getString(index) ?: return@mapNotNull null
              PlayerCapability.fromString(value)
                ?: throw IllegalArgumentException("Invalid notificationCapability value: $value")
            }
          }
        }
    }

    if (map.hasKey("repeatMode")) {
      repeatMode = map.getString("repeatMode")?.let { RepeatMode.fromString(it) } ?: RepeatMode.OFF
    }

    // Android-specific runtime options (all under android.*)
    val androidMap = if (map.hasKey("android")) map.getMap("android") else null
    androidMap?.let { android ->
      if (android.hasKey("ratingType")) {
        ratingType = android.getString("ratingType")?.let { RatingType.fromString(it) }
      }
      if (android.hasKey("appKilledPlaybackBehavior")) {
        appKilledPlaybackBehavior = android.getString("appKilledPlaybackBehavior")
      }
      if (android.hasKey("skipSilence")) {
        skipSilence = android.getBoolean("skipSilence")
      }
      if (android.hasKey("shuffle")) {
        shuffle = android.getBoolean("shuffle")
      }
    }
  }

  fun toBridge(): WritableMap {
    val result = Arguments.createMap()

    // Add jump intervals (always include these core values)
    result.putDouble("forwardJumpInterval", forwardJumpInterval)
    result.putDouble("backwardJumpInterval", backwardJumpInterval)

    // Add progress update interval (always include, null means disabled)
    if (progressUpdateEventInterval != null) {
      result.putDouble("progressUpdateEventInterval", progressUpdateEventInterval)
    } else {
      result.putNull("progressUpdateEventInterval")
    }

    // Add capabilities (always include, even if empty)
    val capabilitiesArray = Arguments.createArray()
    capabilities.forEach { cap -> capabilitiesArray.pushString(cap.string) }
    result.putArray("capabilities", capabilitiesArray)

    // Add notification capabilities if set (null means not set, different from empty)
    notificationCapabilities?.let { caps ->
      val notificationCapsArray = Arguments.createArray()
      caps.forEach { cap -> notificationCapsArray.pushString(cap.string) }
      result.putArray("notificationCapabilities", notificationCapsArray)
    }

    // Add repeat mode (always include)
    result.putString("repeatMode", repeatMode.string)

    // Add Android-specific options if any are set
    val hasAndroidOptions =
      ratingType != null ||
        appKilledPlaybackBehavior != null ||
        skipSilence != null ||
        shuffle != null
    if (hasAndroidOptions) {
      val androidOptions = Arguments.createMap()

      ratingType?.let { androidOptions.putString("ratingType", it.string) }
      appKilledPlaybackBehavior?.let { androidOptions.putString("appKilledPlaybackBehavior", it) }
      skipSilence?.let { androidOptions.putBoolean("skipSilence", it) }
      shuffle?.let { androidOptions.putBoolean("shuffle", it) }

      result.putMap("android", androidOptions)
    }

    return result
  }
}
